crates/wiggle/wasmtime/macro/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+            let mut handle_early_error = quote! { panic!("error: {:?}", e) };
+
+            // The first result is returned bare right now...
+            if let Some(ret) = results.next() {
+                handle_early_error = quote! { return e.into() };
+                match &*ret.tref.type_() {
+                    // Eventually we'll want to add support for more returned
+                    // types, but for now let's just conform to what `*.witx`
+                    // definitions currently use.
+                    witx::Type::Enum(e) => match e.repr {
+                        witx::IntRepr::U16 => {
+                            returns.push(quote! { types::I32 });
+                            ret_ty = quote! { i32 };
+                            cvt_ret = quote! { .into() }
+                        }
+                        other => panic!("unsupported ret enum repr {:?}", other),
+                    },
+                    other => panic!("unsupported first return {:?}", other),
+                }
+            }
+
+            // ... and all remaining results are returned via out-poiners
+            for result in results {
+                let name = format_ident!("{}", result.name.as_str());
+                params.push(quote! { types::I32 });
+                shim_arg_decls.push(quote! { #name: i32 });
+                formats.push(format!("{}={{:#x}}", name));
+                format_args.push(name.clone());
+                hostcall_args.push(quote! { #name });
+            }
+
+            let format_str = format!("{}({})", name, formats.join(", "));
+            ctor_externs.push(quote! {
+                let my_cx = cx.clone();
+                let #name_ident = wasmtime::Func::wrap(
+                    store,
+                    move |caller: wasmtime::Caller<'_> #(,#shim_arg_decls)*| -> #ret_ty {
+                        log::trace!(
+                            #format_str,
+                            #(#format_args),*
+                        );
+                        unsafe {
+                            let mem = match caller.get_export("memory") {
+                                Some(wasmtime::Extern::Memory(m)) => m,
+                                _ => {
+                                    log::warn!("callee does not export a memory as \"memory\"");
+                                    let e = wasi_common::wasi::Errno::Inval;
+                                    #handle_early_error
+                                }
+                            };
+                            // Wiggle does not expose any methods for
+                            // functions to re-enter the WebAssembly module,
+                            // or expose the memory via non-wiggle mechanisms.
+                            // Therefore, creating a new BorrowChecker at the
+                            // root of each function invocation is correct.
+                            let bc = #runtime::BorrowChecker::new();
```

Comment:
Technically this isn't quite true since wasi-common supports virtual handles, right?

Reply:
The `Handle` trait doesn't expose `GuestPtr` or any types which contain one, so I think technically we are ok. All of the interesting uses of `GuestPtr` get translated into `std::io::IoSlice{Mut}` before being passed to the `Handle` methods.

Reply:
I'm not necessarily worried about `GuestPtr`, moreso that if you run arbitrary code it could reenter wasm which could invalidate `GuestPtr` values on the stack previously. The `Handle` type I think can run arbitrary code, right? (and internally it can call something which reenters the original wasm module)

Reply:
Ah, right, I forgot about that aspect. Yes, `Handle` impls can execute whatever code they like, so it is definitely possible for them to re-enter the instance. That means we should be finding some way to provide a `BorrowChecker` that has the same lifetime as the instance, because the lifetime of the call into these functions is no longer sufficient.

I'm not sure how to do that, but I can fix it in this PR if you can outline how I should, or we can leave it for a separate PR.

Reply:
There's a number of ways that we could do this, but I don't think any of them are relatively easy. What we probably want to do is to integrate this into `wasmtime`-the-crate itself, and that way entering into wasm could *always* check to make sure there are no active borrows because we have tight control over the wasm entry points.

Perhaps we could have a discussion in an issue about how to best integrate this tightly with the wasmtime crate itself? I feel like borrowing from guest memory is so common it's justified to put this in the core functionality.

Reply:
https://github.com/bytecodealliance/wasmtime/issues/1917